### PR TITLE
refactor!: checksum operations into pure functions

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
- Change the `algorithm` into a str enum type
- Add support for sha3 and sha256 algos
- Change `Source.load_content` and `Source.compute_checksum` mutating methods on `Source` to instead be pure functions derived from the information in `Source`
NOTE: This is a breaking change, and will require using the `Source` class slightly differently, including updating it with a new version of itself if you wish to store files in a different way than it may have originally specified. Also note that the content will still fetch itself when requested, but it is not no longer being updated in `Source` object (see previous note)
- Fix a bug with hash type sizes in BIP122 URI
- Added a property to verify checksum correctness
- Add a `Hex` type to further validate hex-encoded strings for addresses, blocks, transactions, checksums, and bytecode types
NOTE: This is somewhat a breaking change because now all hex types (except BIP122 URIs) must be prepended w/ `0x` (see changes to OpenZeppelin manifest). Should be interoperable as long as you purge any cached manifests.

fixes: #16
fixes" #18

### How I did it
Heavier use of Pydantic

### How to verify it
Validate changes required to `ape` core

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: ethpm_types/contract_type.py [hunk 1]
```diff
@@ -241,11 +241,13 @@ def validate_uri(cls, uri):
     @classmethod
     def validate_genesis_hash(cls, uri):
         genesis_hash, _, _ = uri.replace("blockchain://", "").split("/")
-        assert is_valid_hash(genesis_hash), f"hash is not valid: {genesis_hash}"
+        assert is_valid_hex("0x" + genesis_hash), f"hash is not valid: {genesis_hash}"
```
Comment (unparalleled-js) [inline ethpm_types/contract_type.py:244]: https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html

File: ethpm_types/source.py [hunk 2]
```diff
@@ -13,41 +15,72 @@ class Compiler(BaseModel):
...
+    """Checksum information about the contents of a source file."""
+
+    algorithm: Algorithm
+    hash: Hex
 
 
 class Source(BaseModel):
+    """Information about a source file included in a Package Manifest."""
+
+    """Array of urls that resolve to the same source file."""
```
Comment (unparalleled-js) [inline ethpm_types/source.py:27]: Are these docs looking okay when built?

I think they actually have to be underneath the property for some weird reason. I have tried to put them above before and concluded that it didn't work.

File: ethpm_types/source.py [hunk 3]
```diff
@@ -13,41 +15,72 @@ class Compiler(BaseModel):
...
         if len(self.urls) == 0:
-            return
+            raise ValueError("No content to fetch.")
 
         response = urllib.request.urlopen(self.urls[0])
-        self.content = response.read().decode("utf-8")
+        content = response.read().decode("utf-8")
+
+        if self.content and self.content != content:
+            raise ValueError("Content mismatches stored.")
```
Comment (unparalleled-js) [inline ethpm_types/source.py:57]: Is mismatches supposed to be "mismatched" here?

File: ethpm_types/source.py [hunk 4]
```diff
@@ -13,41 +15,72 @@ class Compiler(BaseModel):
...
     license: Optional[str] = None
 
-    def load_content(self):
+    def fetch_content(self) -> str:
         """Loads resource at ``urls`` into ``content``."""
         if len(self.urls) == 0:
-            return
+            raise ValueError("No content to fetch.")
 
         response = urllib.request.urlopen(self.urls[0])
```
Comment (unparalleled-js) [inline ethpm_types/source.py:53]: should we maybe verify the status code on this response before continuing?

File: ethpm_types/utils.py [hunk 5]
```diff
@@ -1,20 +1,75 @@
...
+    if set(data[2:].lower()) > set("1234567890abcdef"):
         return False
 
     if len(data) % 2 != 0:
         return False
 
     return True
+
+
+class Hex(str):
```
Comment (unparalleled-js) [inline ethpm_types/utils.py:24]: Would this class be useful outside of ethpm-types?
Like we do need our own "eth-base-types` repository?

Comment (fubuloubu) [reply ethpm_types/utils.py:24]: Maybe eventually

File: ethpm_types/source.py (hunk 4 above)
Comment (fubuloubu) [reply ethpm_types/source.py:53]: Agreed

File: ethpm_types/source.py (hunk 3 above)
Comment (fubuloubu) [reply ethpm_types/source.py:57]:
```suggestion
            raise ValueError("Content mismatched stored value.")
```

File: ethpm_types/contract_type.py (hunk 1 above)
Comment (fubuloubu) [reply ethpm_types/contract_type.py:244]: Yeah should flip these to `if ...: raise` statments

Comment (fubuloubu) [reply ethpm_types/contract_type.py]: 30350a9

File: ethpm_types/source.py (hunk 2 above)
Comment (fubuloubu) [reply ethpm_types/source.py]: 7297d6b

File: ethpm_types/source.py (hunk 3 above)
Comment (fubuloubu) [reply ethpm_types/source.py]: 19657d8

File: ethpm_types/contract_type.py [hunk 6]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
     def validate_uri(cls, uri):
-        assert uri.startswith("blockchain://"), "Must use 'blockchain' protocol"
-        assert (
-            len(uri.replace("blockchain://", "").split("/")) == 3
-        ), "must be referenced via <genesis_hash>/block/<block_hash>"
+        if not uri.startswith("blockchain://"):
+            raise ValueError("Must use 'blockchain' protocol")
+
+        if len(uri.replace("blockchain://", "").split("/")) != 3:
+            raise ValueError("must be referenced via <genesis_hash>/block/<block_hash>")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:238]: Nitpick: standardize the returned error strings
```suggestion
        if not uri.startswith("blockchain://"):
            raise ValueError("Must use 'blockchain' protocol.")

        if len(uri.replace("blockchain://", "").split("/")) != 3:
            raise ValueError("Must be referenced via <genesis_hash>/block/<block_hash>.")
```
or 
```suggestion
        if not uri.startswith("blockchain://"):
            raise ValueError("must use 'blockchain' protocol")

        if len(uri.replace("blockchain://", "").split("/")) != 3:
            raise ValueError("must be referenced via <genesis_hash>/block/<block_hash>")
```

File: ethpm_types/utils.py [hunk 7]
```diff
@@ -1,20 +1,75 @@
...
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> "Hex":
+        return cls("0x" + data.hex())
+
+    def to_bytes(self) -> bytes:
+        return bytes.fromhex(self[2:])
+
+
+def compute_checksum(content: bytes, algorithm: Algorithm = Algorithm.MD5) -> Hex:
```
Comment (evanblank3) [inline ethpm_types/utils.py:61]: Does this need a doc string?

Comment (fubuloubu) [reply ethpm_types/utils.py:61]: idk maybe. it is `utils` lol

**Review by unparalleled-js — APPROVED** [review]

File: ethpm_types/utils.py (hunk 5 above)
Comment (fubuloubu) [reply ethpm_types/utils.py:24]: Okay punting to #18

File: ethpm_types/utils.py (hunk 7 above)
Comment (fubuloubu) [reply ethpm_types/utils.py:61]: I don't see a problem with exposing it either lol

File: ethpm_types/contract_type.py [hunk 8]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
+        if not uri.startswith("blockchain://"):
+            raise ValueError("Must use 'blockchain' protocol.")
+
+        if len(uri.replace("blockchain://", "").split("/")) != 3:
+            raise ValueError("Must be referenced via <genesis_hash>/block/<block_hash>.")
+
         _, block_keyword, _ = uri.replace("blockchain://", "").split("/")
-        assert block_keyword == "block", "must use block reference"
+        if block_keyword != "block":
+            raise ValueError("must use block reference")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:242]:
```suggestion
            raise ValueError("Must use block reference.")
```

File: ethpm_types/contract_type.py [hunk 9]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
+
         return uri
 
     @classmethod
     def validate_genesis_hash(cls, uri):
         genesis_hash, _, _ = uri.replace("blockchain://", "").split("/")
-        assert is_valid_hash(genesis_hash), f"hash is not valid: {genesis_hash}"
+        if not is_valid_hex("0x" + genesis_hash):
+            raise ValueError(f"hash is not valid: {genesis_hash}")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:250]:
```suggestion
            raise ValueError(f"Hash is not valid: {genesis_hash}.")
```

File: ethpm_types/contract_type.py [hunk 10]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
+
+        if len(genesis_hash) != 64:
+            raise ValueError(f"hash is not valid length: {genesis_hash}")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:253]:
```suggestion
            raise ValueError(f"Hash is not valid length: {genesis_hash}.")
```

File: ethpm_types/contract_type.py [hunk 11]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
+
         return uri
 
     @classmethod
     def validate_block_hash(cls, uri):
         _, _, block_hash = uri.replace("blockchain://", "").split("/")
-        assert is_valid_hash(block_hash), f"hash is not valid: {block_hash}"
+        if not is_valid_hex("0x" + block_hash):
+            raise ValueError(f"hash is not valid: {block_hash}")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:261]:
```suggestion
            raise ValueError(f"Hash is not valid: {block_hash}.")
```

File: ethpm_types/contract_type.py [hunk 12]
```diff
@@ -227,25 +227,40 @@ def __modify_schema__(cls, field_schema):
...
+
+        if len(block_hash) != 64:
+            raise ValueError(f"hash is not valid length: {block_hash}")
```
Comment (evanblank3) [inline ethpm_types/contract_type.py:264]:
```suggestion
            raise ValueError(f"Hash is not valid length: {block_hash}.")
```

**Review by evanblank3 — APPROVED** [review]
Nitpick, feel free to ignore.

**Comment by fubuloubu** [comment]
This is breaking change, need to coordinate upgrade w/ ape core

**Review by NotPeopling2day — APPROVED** [review]

**Pull request merged**